### PR TITLE
[ASP.NET Core] Enable diagnostic source events using explicit names

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Towards #3495 .

## Changes

This change will enable diagnostic source event subscription for asp.net core for the specified list of events.

ASP.NET Core library is using the [best practices](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md#filtering) defined to write diagnostic source events.

Reference:
https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/MvcCoreDiagnosticListenerExtensions.cs
https://github.com/dotnet/aspnetcore/blob/main/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

This may also help with some perf as it will prevent the payload creation and writing of non-desired events.

Also, this can be extended to other instrumentation libraries based on DiagnosticSource events.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs [hunk 1]
```diff
@@ -23,11 +24,23 @@ namespace OpenTelemetry.Instrumentation.AspNetCore
...
+
+        private readonly Func<string, object, object, bool> isEnabled = (eventName, obj1, obj2)
+            => DiagnosticSourceEvents.Contains(eventName);
+
         private readonly DiagnosticSourceSubscriber diagnosticSourceSubscriber;
 
         public AspNetCoreInstrumentation(HttpInListener httpInListener)
         {
-            this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(httpInListener, null);
+            this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(httpInListener, this.isEnabled);
```
Comment (alanwest) [inline src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:43]: 👍 I think using an `isEnabled` delegate to define the events we're interested in is a good step in the right direction.

If we were to do this to all our instrumentation then in the end I think we'd still want to do away with pattern matching here https://github.com/open-telemetry/opentelemetry-dotnet/blob/27658f058cc248500805d66ae2ab65c2099d6400/src/OpenTelemetry/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs#L43.

That is, I no longer see the value this really brings. I think it makes more sense to change [ListenerHandler](https://github.com/open-telemetry/opentelemetry-dotnet/blob/27658f058cc248500805d66ae2ab65c2099d6400/src/OpenTelemetry/DiagnosticSourceInstrumentation/ListenerHandler.cs#L23) to just have one method like the following and have it switch on event name - much like `OnCustom` does today.

```csharp
public virtual void OnEventWritten(Activity activity, string eventName, object payload)
```

All that said, I think we could do this in stages. So leave `ListenerHandler` alone for now. Do what you've done here to define the events we want. Then come back later and decide whether we want to refactor `ListenerHander` and friends.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:43]: good that all these DS helpers are kept internal. we can break them apart.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:43]: even Activity is not required to be part of this OnEventWritten. It gives wrong information that Activity is somehow part of the event. It is not. It is part of ambient context, and anyone interested can use Activity.Current to get it. Or use other mechanism (like from inside httpcontext somewhere etc) if Activity.Current is not reliable.

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:43]: I will do a follow up PR for moving all events to custom callback.

**Review by utpilla — APPROVED** [review]

File: test/Benchmarks/Instrumentation/AspNetCoreInstrumentationBenchmarks.cs [hunk 2]
```diff
@@ -27,17 +27,17 @@
...
+  [Host] : .NET 6.0.9 (6.0.922.41905), X64 RyuJIT
 
 Job=InProcess  Toolchain=InProcessEmitToolchain
 
 |                                      Method |     Mean |   Error |  StdDev |  Gen 0 | Allocated |
 |-------------------------------------------- |---------:|--------:|--------:|-------:|----------:|
-|                 UninstrumentedAspNetCoreApp | 155.6 us | 2.63 us | 2.33 us | 0.9766 |      5 KB |
-| InstrumentedAspNetCoreAppWithDefaultOptions | 176.8 us | 3.24 us | 2.70 us | 1.2207 |      7 KB |
+|                 UninstrumentedAspNetCoreApp | 164.7 us | 1.66 us | 1.39 us | 0.9766 |      5 KB |
+| InstrumentedAspNetCoreAppWithDefaultOptions | 178.7 us | 2.65 us | 2.35 us | 0.9766 |      5 KB |
```
Comment (cijothomas) [inline test/Benchmarks/Instrumentation/AspNetCoreInstrumentationBenchmarks.cs:40]: :100: ❤️

**Review by cijothomas — APPROVED** [review]
LGTM.
Might be good to add a entry in changelog to highlight the improved performance to motivate ppl to update to newer version. We usually do not add changelog for internal only changes, but the perf gain here should be worth mentioning.

File: src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md [hunk 3]
```diff
@@ -2,15 +2,18 @@
 
 ## Unreleased
 
-* Fix issue where when an application has an ExceptionFilter, the exception data
-  wouldn't be collected.
-  ([#3475](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3475))
+* Updated DiagnosticSource event subscription to specific set of events.
```
Comment (utpilla) [inline src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:5]:
```suggestion
* AspNetCore instrumentation causes no extra allocation now as we have now updated DiagnosticSource event subscription to a specific set of events.
```

Comment (utpilla) [reply src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md:5]: > AspNetCore instrumentation causes no extra allocation

^ Ony if this is factually correct

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/CHANGELOG.md]: There is some extra allocation - it is not exactly 0. I will try with `params` to see if we can show the difference.

**Review by utpilla — APPROVED** [review]

**Comment by CodeBlanch** [comment]
@vishweshbankwar Hey I think this PR might have broken stuff. I'm testing the AspNetCore instrumentation using the example app. Not seeing _any_ spans when running in net6.0. Took me a while but I got here:

image

Looks like a mismatch from what is being used vs what is defined?

**Pull request merged**